akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -283,14 +291,21 @@ private[akka] object StreamLayout {
 
     override def toString =
       s"""
-        | Module: ${this.attributes.nameOption.getOrElse("unnamed")}
-        | Modules: ${subModules.toSeq.map(m ⇒ "   " + m.attributes.nameOption.getOrElse(m.getClass.getName)).mkString("\n")}
+        | Module: ${this.attributes.nameOrDefault("unnamed")}
+        | Modules: ${subModules.iterator.map(m ⇒ "   " + m.attributes.nameOrDefault(m.getClass.getName)).mkString("\n")}
         | Downstreams:
-        | ${downstreams.map { case (in, out) ⇒ s"   $in -> $out" }.mkString("\n")}
+        | ${downstreams.iterator.map { case (in, out) ⇒ s"   $in -> $out" }.mkString("\n")}
         | Upstreams:
-        | ${upstreams.map { case (out, in) ⇒ s"   $out -> $in" }.mkString("\n")}
+        | ${upstreams.iterator.map { case (out, in) ⇒ s"   $out -> $in" }.mkString("\n")}
       """.stripMargin
 
+    override def describe: String =
+      attributes.nameOrDefault(null) match {
+        case null ⇒ "(" + subModules.iterator.map(_.describe).mkString(" ~> ") + ")"
```

Comment:
@drewhk Sadly, since subModules is a Set, there's no way to get stable ordering of operations. Any ideas how to achieve that?

~~~scala
scala> res3 via res5
res6: akka.stream.scaladsl.Source[String,Unit] = fold ~> futureSource ~> mapAsyncUnordered ~> filter ~> map

scala> res3 via res5
res7: akka.stream.scaladsl.Source[String,Unit] = filter ~> fold ~> futureSource ~> mapAsyncUnordered ~> map

scala> res3 via res5
res8: akka.stream.scaladsl.Source[String,Unit] = filter ~> fold ~> futureSource ~> mapAsyncUnordered ~> map

scala> res3 via res5
res9: akka.stream.scaladsl.Source[String,Unit] = fold ~> futureSource ~> filter ~> mapAsyncUnordered ~> map
~~~